src/components/ec/cuda/ec_cuda_executor_interruptible.c

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020-2022.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "ec_cuda_executor.h"
+#include "components/mc/ucc_mc.h"
+#include "utils/ucc_atomic.h"
+
+ucc_status_t ucc_cuda_executor_interruptible_get_stream(cudaStream_t *stream)
+{
+    static uint32_t last_used = 0;
+    cudaError_t     cuda_st;
+    int             i, j;
+    uint32_t        id;
+
+    if (ucc_unlikely(ucc_ec_cuda.exec_streams[0] == NULL)) {
+        ucc_spin_lock(&ucc_ec_cuda.init_spinlock);
+        if (ucc_ec_cuda.exec_streams[0] != NULL) {
+            goto unlock;
+        }
+
+        for(i = 0; i < EC_CUDA_CONFIG->exec_num_streams; i++) {
+            cuda_st = cudaStreamCreateWithFlags(&ucc_ec_cuda.exec_streams[i],
```

Comment:
@Sergei-Lebedev i've discovered a race in MC_INIT in master (Fix: https://github.com/openucx/ucc/pull/468/commits/fd1b5529ece38de433b039caf029be5a96aa01a9). Happens in MT case. Apparently, internal CUDA implementation only guarantees that the stream ptr can be safely re-used once the corresponding cudaStreamCreateWithFlags returns. The race is: one thread gets the lock and calls streamCreate: inside the call the shared stream ptr gets non Null value, but stream is not entirely initialized yet. At the same time another thread seeys stream != NULL ang goes to cudaMemcpy -> this causes an error from cuda: [1650316036.432440] [swx-clx02:15445:0]         mc_cuda.c:309  cuda mc ERROR failed to launch cudaMemcpyAsync,  dst 0x7f441bf00000, src 0x7f441c001240, len 4 cuda error 400(invalid resource handle)

The fix is to use local variable and atomic_cswap. In the case of array i think you can do the same but check the last ptr of the array for NULL (or loop backward)

Reply:
good catch! fixed